### PR TITLE
iio: ability to set IIO context attributes from application code

### PR DESCRIPTION
**Pull request by mphalke**
Implemented the changes in no-os IIO library to allow user to set the context attributes and values
from an application code.

Signed-off-by: mahphalke <Mahesh.Phalke@analog.com>

### CODE REVIEW

File: iio/iio.c [hunk 1]
```diff
@@ -1141,7 +1141,22 @@ static uint32_t iio_generate_device_xml(struct iio_device *device, char *name,
...
 	i = 0;
-	i += snprintf(buff, max(n - i, 0),
+
+	/* Write context attributes */
+	if (device->context_attributes)
+		for (j = 0; device->context_attributes[j].name; j++) {
+			i += snprintf(buff + i,
+				      max(n - i, 0),
+				      "<context-attribute name=\"%s\" ",
+				      device->context_attributes[j].name);
```
Comment (buha) [inline iio/iio.c:1151]: I'm wondering if it'd make sense to put some default values for .name and .value but I guess it's also fine without any defaults

Comment (mphalke) [reply iio/iio.c:1151]: @buha - sorry, is this question for me or other reviewers? Where do you suggest default values, I didn't get you?

Comment (buha) [reply iio/iio.c:1151]: I was just thinking out loud. I think currently the name defaults to "no-OS" and the value is the git version that defaults to something like "1.0.0-g0000000". If device->context_attributes[j].name is NULL, I think this is undefined behaviour in C:
https://stackoverflow.com/questions/11589342/what-is-the-behavior-of-printing-null-with-printfs-s-specifier

So maybe just initialize device->context_attributes[j].name with "no-OS" if it is NULL, prior to the snprintf call.

Comment (mphalke) [reply iio/iio.c:1151]: Looks like NULL check done in the for loop. Would that be good?
```
for (j = 0; device->context_attributes[j].name; j++)
```

**Review by adrimbarean — APPROVED** [review]
Looks good to me.

**Comment by mphalke** [comment]
Hi,
Are we good with this PR?

**Review by amiclaus — APPROVED** [review]

File: iio/iio.c (hunk 1 above)
Comment (buha) [reply iio/iio.c:1151]: seems ok then, just make sure to fix conflicts

**Comment by mphalke** [comment]
Not exactly sure from where the conflict is coming from. Any clue?
Do I need to merge the latest changes?

**Review by amiclaus — COMMENTED** [review]
image

`max` function was renamed in the meantime to `no_os_max`
https://github.com/analogdevicesinc/no-OS/blob/a56c44f093f2764f9350dc1120e96425dde01d88/include/no_os_util.h#L68

**Review by amiclaus — COMMENTED** [review]
> Not exactly sure from where the conflict is coming from. Any clue? Do I need to merge the latest changes?

So you will have to rebase your branch with latest master branch. And rename in your patch the `max` macro.

**Comment by mphalke** [comment]
> > Not exactly sure from where the conflict is coming from. Any clue? Do I need to merge the latest changes?
> 
> So you will have to rebase your branch with latest master branch. And rename in your patch the `max` macro.

Thanks. I'll rebase with master and make this change.

File: iio/iio.c [hunk 2]
```diff
@@ -1143,7 +1143,22 @@ static uint32_t iio_generate_device_xml(struct iio_device *device, char *name,
 		buff = ch_id;
 
 	i = 0;
-	i += snprintf(buff, no_os_max(n - i, 0),
+
+	/* Write context attributes */
+	if (device->context_attributes)
+		for (j = 0; device->context_attributes[j].name; j++) {
+			i += snprintf(buff + i,
+				      max(n - i, 0),
```
Comment (rbolboac) [inline iio/iio.c:1151]: Please use no_os_max instead of max.

**Review by rbolboac — CHANGES_REQUESTED** [review]
Besides these 2 small remarks, lgtm.

**Comment by mphalke** [comment]
V1.0:

- Replaced max() with no_os_max() in iio.c file

**Review by rbolboac — APPROVED** [review]
lgtm

**Review by amiclaus — APPROVED** [review]

**Pull request merged**